File: product-of-array-except-self-238/product_of_array_except_self.rb
# @param {Integer[]} nums
# @return {Integer[]}
def product_except_self(nums)
    l = nums.length
    result = [1]*l

    pre,post=1,1

    i = 0
    until i == l do
        result[i] = pre
        pre = pre * nums[i]
        i+=1
    end

    i = l-1
    until i < 0 do
        result[i] = result[i] * post
        post = post * nums[i]
        i-=1
    end

    return result
end

File: product-of-array-except-self-238/product_of_array_except_self.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function(nums) {
    const l = nums.length;
    const result = Array(l).fill(l);
    
    let pre=1, post = 1;

    for(let i=0; i<l; i++) {
        result[i] = pre;
        pre = pre * nums[i];
    }

    for(let i=l-1; i>=0; i--) {
        result[i] = result[i] * post;
        post = post * nums[i];
    }

    return result;
};